feat: Open layers in a read only mode with Qgis::ProjectReadFlag::ForceReadOnlyLayers

rldhont opened the pull request:
## Description

In most cases of use of QGIS Server, it is not necessary to access the layers in write mode.
The read-only mode is sufficient.

We would like to introduce a new flag `Qgis::ProjectReadFlag::ForceLayerReadOnly` to
open layers in a read-only mode when the project is read. To activate this flag for QGIS Server,
the environmental variable `QGIS_SERVER_FORCE_READONLY_LAYERS` can be defined.

In addition with `Qgis::ProjectReadFlag::ForceLayerReadOnly`, new flags are available for reading 
layer XML `QgsMapLayer::FlagForceReadOnly` and for dataprovider contruction `QgsDataProvider::ForceReadOnly`.

The `QgsVectorLayer::LayerOptions` has been extended by adding a new property `forceReadOnly` to programmatically open a layer in read only mode.

Funded by Ifremer https://wwz.ifremer.fr/

rldhont committed:
feat: Open layers in a read only mode with Qgis::ProjectReadFlag::ForceLayerReadOnly

In most cases of use of QGIS Server, it is not necessary to access the layers in write mode.
The read-only mode is sufficient.

We would like to introduce a new flag Qgis::ProjectReadFlag::ForceLayerReadOnly to
open layers in a read-only mode.

nyalldawson commented on python/core/auto_generated/qgis.sip.in:
```diff
@@ -1175,6 +1175,7 @@ The development version
       DontStoreOriginalStyles,
       DontLoad3DViews,
       DontLoadProjectStyles,
+      ForceLayerReadOnly,
```
```suggestion
      ForceReadOnlyLayers,
```
??

nyalldawson commented on src/core/providers/qgsdataprovider.h:
```diff
@@ -125,6 +125,7 @@ class CORE_EXPORT QgsDataProvider : public QObject
       FlagLoadDefaultStyle = 1 << 2, //!< Reset the layer's style to the default for the datasource
       SkipGetExtent = 1 << 3, //!< Skip the extent from provider
       SkipFullScan = 1 << 4, //!< Skip expensive full scan on files (i.e. on delimited text) (since QGIS 3.24)
+      SkipGetEditable = 1 << 5, //!< Skip the edit capabilities (since QGIS 3.28)
```
This wording is not very nice -- can we go with just FlagForceReadOnly instead?

nyalldawson commented on src/core/qgis.h:
```diff
@@ -2001,6 +2001,7 @@ class CORE_EXPORT Qgis
       DontStoreOriginalStyles SIP_MONKEYPATCH_COMPAT_NAME( FlagDontStoreOriginalStyles ) = 1 << 3, //!< Skip the initial XML style storage for layers. Useful for minimising project load times in non-interactive contexts.
       DontLoad3DViews SIP_MONKEYPATCH_COMPAT_NAME( FlagDontLoad3DViews ) = 1 << 4, //!< Skip loading 3D views (since QGIS 3.26)
       DontLoadProjectStyles = 1 << 5, //!< Skip loading project style databases (deprecated -- use ProjectCapability::ProjectStyles flag instead)
+      ForceLayerReadOnly SIP_MONKEYPATCH_COMPAT_NAME( FlagForceLayerReadOnly ) = 1 << 6, //!< Open layers in a read-only mode. (since QGIS 3.28)
```
```suggestion
      ForceLayerReadOnly = 1 << 6, //!< Open layers in a read-only mode. (since QGIS 3.28)
```

We only need the compat_name when maintaining existing API is a factor -- here it's a completely new flag so isn't required.

nyalldawson reviewed (commented):
Thanks for tackling this @rldhont, I agree it's a much needed addition!

I gave an earlier attempt to this at https://github.com/nyalldawson/QGIS/commit/e8d15e1bb89f740209ba3b3ac1e169cb62e26796, which didn't get quite as far as what you've done here :+1: There's a few parts from my earlier effort which I think need incorporating here though:

1. We need a way to construct an entirely new layer with the read only flag -- so the current approach of adding the read only setting only as a new QgsMapLayer::ReadFlag doesn't allow this. Specifically we need to be able to create layers for situations where we know they are always read only, such as to show the world map preview in the crs selector widget. In this case we never read existing layer properties and we need the read flag accessible directly in the QgsMapLayer constructor. (I attempted to address this by adding the flag as an option to the various Qgs*Layer::LayerOption classes)
2. Whenever we load the world map file we should use the read only flag (solves https://github.com/qgis/QGIS/issues/35383)
3. The read only flag needs to get cloned whenever a layer/data provider is cloned too

rldhont commented:
Thanks @nyalldawson for the review. I have integrated the requested changes.

I will take a look at your suggestions.

rldhont committed:
feat: New force read only option for vector layer

To be able to open vector layer in a read-only mode, a new vector layer option has been defined to do it.

This option is used to open world_map.gpkg and fixes #35383

rldhont commented:
Hi @nyalldawson, I think I have made all the changes requested. QgsVectorLayer has already a mReadOnly private properties. 

All other layers and providers are open in a read-only mode by default.

nyalldawson commented:
@rldhont 
This is looking good!

One thing I'm a bit unsure about is whether it's safe to reuse the existing vector layer mReadOnly flag for this purpose. Looking into where that flag is currently set, there's this bit of code in QgsGeometryChecker:

```
      it.value()->layer()->setReadOnly( true );
      // Enter update mode to defer ogr dataset repacking until the checker has finished
      it.value()->layer()->dataProvider()->enterUpdateMode();
```

I don't fully understand the rationale behind that code, but it's likely going to break if we make the QgsVectorLayer::setReadOnly flag force the data provider into a read only mode too.

It might be safest to just use a new member to store the "data provider is read only" flag and avoid a subtle change in the existing api logic (where setReadOnly only hides editing capability from the QGIS UI, it doesn't actually force the layer into a true read only mode)

rldhont commented:
Hi @nyalldawson, I have updated the way to store `LayerOptions.forceReadOnly`. Do we have to store it in the layer XML or in the Layer datasource ? and How ?

andreasneumann commented:
@rldhont - do you expect a performance improvement with these read-only layers, or do you implement them for a different purpose?

And if there is an expected performance improvement - would this be noticeable primarily at project load time, or for other operations as well?

mdouchin commented:
Hi @andreasneumann We indeed expect a performance improvement for database layers (PostgreSQL and Oracle) because fewer queries will be made when initializing the layers during the project read.

Once the project is ready (and in QGIS Server cache for example), I think it won't change performance for other operations on PostgreSQL/Oracle layers. But definitely yes during the project loading.

I think it will also help to handle GeoPackage layers from QGIS Server, since no side files will be created. For this format, concurrent reading performance (multiple requests on the same GeoPackage) may also be improved, also for GetMap and other layer based queries.

Hopefully we still have performance tests across QGIS Server versions to check this performance difference before and after this PR.

rldhont commented:
Hi @andreasneumann, we expect a performance improvement for project loading with OGR, Postgres and Oracle layers. I hope that force read only for OGR layer is faster than trying to open in read write mode. I think that for QGIS server in most cases, it is not needed to open layers write capabilities.

rldhont commented:
All tests are green. I have updated the PR description for documentation.

@nyalldawson do you have any change requests ?

Is anyone else likely to do a code review?

rldhont commented:
I have rebased to avoid conflict.

nyalldawson commented on python/server/auto_generated/qgsserversettings.sip.in:
```diff
@@ -278,6 +279,16 @@ The default value is ``False``, this value can be changed by setting the environ
 .. versionadded:: 3.16
+%End
+
+    bool ForceReadOnlyLayers() const;
```
```suggestion
    bool forceReadOnlyLayers() const;
```

rldhont commented on src/core/raster/qgsrasterlayer.cpp:
```diff
@@ -2069,6 +2069,10 @@ bool QgsRasterLayer::readXml( const QDomNode &layer_node, QgsReadWriteContext &c
     {
       flags |= QgsDataProvider::FlagTrustDataSource;
     }
+    if ( mReadFlags & QgsMapLayer::FlagForceReadOnly )
+    {
+      flags |= QgsDataProvider::ForceReadOnly;
+    }
```
@nyalldawson for others I didn't see write capabilities, but I can add the flag.

elpaso commented on src/core/providers/qgsdataprovider.h:
```diff
@@ -125,6 +125,7 @@ class CORE_EXPORT QgsDataProvider : public QObject
       FlagLoadDefaultStyle = 1 << 2, //!< Reset the layer's style to the default for the datasource
       SkipGetExtent = 1 << 3, //!< Skip the extent from provider
       SkipFullScan = 1 << 4, //!< Skip expensive full scan on files (i.e. on delimited text) (since QGIS 3.24)
+      ForceReadOnly = 1 << 5, //!< Skip the edit capabilities (since QGIS 3.28)
```
```suggestion
      ForceReadOnly = 1 << 5, //!<  Open layer in a read-only mode (since QGIS 3.28)
```

elpaso commented on src/core/vector/qgsvectorlayer.h:
```diff
@@ -2893,6 +2905,9 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
     QgsActionManager *mActions = nullptr;
 
+    //! Flag indicating whether the data source is in read-only mode (editing disabled) or not
+    bool mDSReadOnly = false;
```
Can you please expand this variable name to something human intelligible and add some more explanation about what is the difference from `mReadOnly` and how they relate?
```suggestion
    bool mDataSourceReadOnly = false;
```

elpaso commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -1520,7 +1520,9 @@ bool QgsPostgresProvider::hasSufficientPermsAndCapabilities()
-    bool inRecovery = false;
+    // Do not check the editable capabilities if the flag has been set
+    // or if the database crashed or connected to a standby server
+    bool inRecovery = ( mReadFlags & QgsDataProvider::ForceReadOnly );
```
I think that you are abusing `inRecovery` here: either use a separate bool or rename it to indicate that the variable carries a different meaning than the status of the recovery.

elpaso reviewed (commented):
General implementation looks good to me, most welcome feature!

elpaso commented:
> Thanks @elpaso for the review, I have made the requested changes.

Thanks, much clearer now.
